### PR TITLE
Feature/add product price and date time to home page

### PR DESCRIPTION
**Pull request by sumantkadam**

### CODE REVIEW

File: ozpricechecker/pricefinderapp/tests/models/test_product.py [hunk 1]
```diff
@@ -18,6 +20,20 @@ def test_str_no_name(self):
     def test_str_name(self):
         product = Product.objects.create(store=self.store, prod_url='/1234/Tomatoes', name='Super Tomatoes')
         self.assertEqual(str(product), 'Woolworths - Super Tomatoes')
+    
+    def test_latest_prod_price_date(self):
+        product = Product.objects.create(store=self.store, prod_url='/1234/Tomatoes', name='Super Tomatoes')
+        productPrice1 = ProductPrice.objects.create(product=product, price=10)
+        time.sleep(1)
```
Comment (ericziethen) [inline ozpricechecker/pricefinderapp/tests/models/test_product.py:30]: Don't use sleep, set the time directly instead

File: ozpricechecker/pricefinderapp/tests/models/test_product.py [hunk 2]
```diff
@@ -18,6 +20,20 @@ def test_str_no_name(self):
     def test_str_name(self):
         product = Product.objects.create(store=self.store, prod_url='/1234/Tomatoes', name='Super Tomatoes')
         self.assertEqual(str(product), 'Woolworths - Super Tomatoes')
+    
+    def test_latest_prod_price_date(self):
```
Comment (ericziethen) [inline ozpricechecker/pricefinderapp/tests/models/test_product.py:24]: It might be easier to maintain if we separate the tests for the 2 properties and have at least a test for each

File: ozpricechecker/pricefinderapp/tests/models/test_product.py [hunk 3]
```diff
@@ -18,6 +20,20 @@ def test_str_no_name(self):
...
+        product = Product.objects.create(store=self.store, prod_url='/1234/Tomatoes', name='Super Tomatoes')
```
Comment (ericziethen) [inline ozpricechecker/pricefinderapp/tests/models/test_product.py:25]: Can we add a test for accessing the Properties when ProductPrice table is empty

File: ozpricechecker/pricefinderapp/tests/models/test_product.py [hunk 4]
```diff
@@ -18,6 +20,20 @@ def test_str_no_name(self):
...
+        time.sleep(1)
+        productPrice2 = ProductPrice.objects.create(product=product, price=50)
+        time.sleep(1)
+        productPrice3 = ProductPrice.objects.create(product=product, price=25)
+        self.assertEqual(product.latest_price, productPrice3.price)
+        self.assertEqual(product.date_for_latest_price, productPrice3.date_time)
+
+        productPrice3.delete()
+        self.assertEqual(product.latest_price, productPrice2.price)
+        self.assertEqual(product.date_for_latest_price, productPrice2.date_time)
```
Comment (ericziethen) [inline ozpricechecker/pricefinderapp/tests/models/test_product.py]: Can we add a test where the latest date for a product doesn't have a price. Since we are currently still supporting this (and might depending on how we decide) we should still make sure we retrieve the latest price.

File: ozpricechecker/templates/pricefinderapp/userproduct_list.html [hunk 5]
```diff
@@ -1,32 +1,32 @@
...
-                    <th scope="col">Threshhold Price</th>
-                    <th scope="col">Action</th>
+                    <th scope="col">
+                        User Product
+                        <a class="badge badge-primary" href="{% url "userproduct_add" %}">+</a>
+                    </th>
+                    <th class="text-right" scope="col">Threshhold Price</th>
+                    <th class="text-right" scope="col">Price</th>
+                    <th scope="col">Curency</th>
+                    <th class="text-center" scope="col">DateTime</th>
```
Comment (ericziethen) [inline ozpricechecker/templates/pricefinderapp/userproduct_list.html]: Can we name that a bit clearer to highlight the fact that it's the Product Price?

**Review by ericziethen — CHANGES_REQUESTED** [review]

File: ozpricechecker/pricefinderapp/tests/models/test_product.py (hunk 1 above)
Comment (ericziethen) [reply ozpricechecker/pricefinderapp/tests/models/test_product.py:30]: Agreed once we allow setting the Data manually on Product Price we need to remove the sleep statements.

File: ozpricechecker/pricefinderapp/tests/models/test_product.py (hunk 2 above)
Comment (ericziethen) [reply ozpricechecker/pricefinderapp/tests/models/test_product.py:24]: Refactor later

File: ozpricechecker/pricefinderapp/tests/models/test_product.py (hunk 3 above)
Comment (ericziethen) [reply ozpricechecker/pricefinderapp/tests/models/test_product.py:25]: Add now if an exception caused otherwise later

File: ozpricechecker/pricefinderapp/tests/models/test_product.py (hunk 4 above)
Comment (ericziethen) [reply ozpricechecker/pricefinderapp/tests/models/test_product.py]: Not needed since we will wont allow this to happen

File: ozpricechecker/templates/pricefinderapp/userproduct_list.html (hunk 5 above)
Comment (sumantkadam) [reply ozpricechecker/templates/pricefinderapp/userproduct_list.html]: Changed name to "Price Date"

File: ozpricechecker/pricefinderapp/tests/models/test_product.py (hunk 3 above)
Comment (sumantkadam) [reply ozpricechecker/pricefinderapp/tests/models/test_product.py:25]: Fixed exception.

File: ozpricechecker/pricefinderapp/models.py [hunk 6]
```diff
@@ -50,6 +50,22 @@ def full_url(self):
             base_url += '/'
         return urljoin(base_url, self.prod_url)
 
+    @property
+    def latest_price(self):
+        """Get the full url for this product."""
+        product_price = ProductPrice.objects.filter(product=self).order_by('-date_time').first()
+        if product_price:
+            return product_price.price
+        return None
```
Comment (ericziethen) [inline ozpricechecker/pricefinderapp/models.py:59]: first() already returns None so the 4 lines can become a single line.
Basically the above is checking for None and then returning None if it found None

File: ozpricechecker/pricefinderapp/models.py [hunk 7]
```diff
@@ -50,6 +50,22 @@ def full_url(self):
...
+
+    @property
+    def date_for_latest_price(self):
+        """Get the full url for this product."""
+        product_price = ProductPrice.objects.filter(product=self).order_by('-date_time').first()
+        if product_price:
+            return product_price.date_time
+        return None
```
Comment (ericziethen) [inline ozpricechecker/pricefinderapp/models.py:67]: Same comment as for latest_price, simplify return statement.

**Review by ericziethen — CHANGES_REQUESTED** [review]

File: ozpricechecker/pricefinderapp/models.py (hunk 6 above)
Comment (sumantkadam) [reply ozpricechecker/pricefinderapp/models.py:59]: @ericziethen 
Here need to return only price only not entire record then how can it be done in single line?

**Review by ericziethen — DISMISSED** [review]

**Review by ericziethen — APPROVED** [review]

**Review by ericziethen — DISMISSED** [review]

**Review by ericziethen — APPROVED** [review]
